### PR TITLE
Read Outputs from Job Server

### PR DESCRIPTION
**Pull request by ghickman**
This adds the ability to create reports (et al) from outputs on job-server, whether they are published or not.  It mirrors much of the GitHub mechanism (a client class and a report class), to handle remote IO and caching, in an effort to keep their respective APIs as similar as possible.

I've set up GitHub and job-server sections in the admin UI:

`Report.clean()` checks that only one section has been completed (ignoring `branch` since it has a default).

⚠️ **Blocked by:** ⚠️
* opensafely-core/osgithub#44
* opensafely-core/job-server#1742

### CODE REVIEW

File: reports/admin.py [hunk 1]
```diff
@@ -24,8 +24,24 @@ class ReportAdmin(admin.ModelAdmin):
...
+            },
+        ),
+        (
+            "Report file details (Jobs site)",
+            {
+                "description": (
+                    "A report file can be hosted on either GitHub or the Jobs site, "
+                    "paste the direct URL to your output on the Jobs site."
+                ),
+                "fields": ["job_server_url"],
```
Comment (rebkwok) [inline reports/admin.py:43]: At some point I assume we'll plan to deal with more than just a single HTML file as a report, but for now would it make sense to say that the url is expected to point to a single HTML file?  The different wording implies that maybe a report has to be an HTML file if it's on GitHub, but maybe could be something else if on the Jobs site

**Review by rebkwok — COMMENTED** [review]
Some questions about the last_updated field and caching

File: reports/models.py [hunk 2]
```diff
@@ -186,8 +197,37 @@ def clean(self):
...
+        # filled in, not all filled in) and also mutually exclusive.
+
+        # Note self.branch has a default so we ignore it for simplicity,
+        # otherwise we'll have to get the default from self._meta.fields
+        github_fields = [self.repo, self.report_html_file_path]
+        all_github = all(f != "" for f in github_fields)
+        some_github = any(f != "" for f in github_fields)
+        no_github = all(f == "" for f in github_fields)
+
+        # no_job_server = all(f == "" for f in job_server_fields)
```
[the remark this answers is not in the document]
Comment (ghickman) [reply reports/models.py:211]: oops, yep 👍

File: reports/views.py [hunk 3]
```diff
@@ -34,13 +36,14 @@ def report_view(request, slug):
         )
         return redirect(report.get_absolute_url())
 
-    github_report = GithubReport(report)
+    remote_cls = GithubReport if report.uses_github else JobServerReport
+    html = process_html(remote_cls(report).get_html())
```
[the remark this answers is not in the document]
Comment (rebkwok) [reply reports/views.py:40]: I think so...although depending on whether you need to do make sure `last_updated` is called separately, you might need to instantiate the `remote_cls` anyway, in which case  you could just send the instance in the context?

[the remark this answers is not in the document]
Comment (ghickman) [reply reports/views.py:40]: Oh yes, lets keep the API consistent between the two backends and have them both call `last_updated` separately from instances 👍

**Comment by ghickman** [comment]
@rebkwok – updated to re-instate the caching, could you take another look?

**Review by rebkwok — APPROVED** [review]

**Pull request merged**